crates/lib3h/src/error.rs

```diff
@@ -0,0 +1,125 @@
+//! Lib3h custom error definition.
+
+use crate::transport::error::TransportError;
+use lib3h_crypto_api::CryptoError;
+use lib3h_protocol::error::{ErrorKind as Lib3hProtocolErrorKind, Lib3hProtocolError};
+use rmp_serde::decode::Error as RMPSerdeDecodeError;
+use std::{error::Error as StdError, fmt, io, result};
+
+/// A type alias for `Result<T, Lib3hError>`.
+pub type Lib3hResult<T> = result::Result<T, Lib3hError>;
+
+/// An error that can occur when interacting with the algorithm.
+#[derive(Debug)]
+pub struct Lib3hError(Box<ErrorKind>);
+
+impl Lib3hError {
+    /// A constructor for `Lib3hError`.
+    pub fn new(kind: ErrorKind) -> Lib3hError {
+        Lib3hError(Box::new(kind))
+    }
+
+    /// Return the specific type of this error.
+    pub fn kind(&self) -> &ErrorKind {
+        &self.0
+    }
+
+    /// Unwrap this error into its underlying type.
+    pub fn into_kind(self) -> ErrorKind {
+        *self.0
+    }
+}
+
+/// The specific type of an error.
+#[derive(Debug)]
+pub enum ErrorKind {
+    /// An I/O error that occurred while processing a data stream.
+    Io(io::Error),
+    /// Error occuring when using `transport`.
+    TransportError(TransportError),
+    /// Error originating from [lib3h_protocol] crate.
+    Lib3hProtocolError(Lib3hProtocolError),
```

Comment:
I believe you got this backwards.
`Lib3hProtocolError` will most probably not have its own errors since its just an api (declarations only).
So I see `Lib3hError` as an ErrorKind of `Lib3hProtocolError` and not the other way around.

Reply:
I understand and this is what I tried first but it gives us circular dependency trouble, as `Lib3h` already depends from `Lib3hProtocol` so I cannot make `Lib3hProtocol` depends from `Lib3h` as well :/

This gave me a little trouble and this is the solution I found. But I'm of course open to discuss it further :)